<a href="https://colab.research.google.com/github/agu3/ds4bmeTest/blob/master/a8_q3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Neural Network Cross Validation


In [0]:
import numpy as np
from keras import models
from keras import layers
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_validate
from sklearn.datasets import make_classification
import pandas as pd
import tensorflow as tf

In [0]:
np.random.seed(0)

In [0]:
! if [ ! -e oasis.csv ];\
  then wget https://raw.githubusercontent.com/bcaffo/ds4bme_intro/master/data/oasis.csv; \
fi;

In [0]:
oasis = pd.read_csv("oasis.csv")
oasis.head()

FLAIR        PD        T1  ...     T1_20     T2_20  GOLD_Lesions
0  1.143692  1.586219 -0.799859  ...  0.219136  0.298662             0
1  1.652552  1.766672 -1.250992  ...  0.061573  0.280972             0
2  1.036099  0.262042 -0.858565  ...  0.020673 -0.259914             0
3  1.037692  0.011104 -1.228796  ... -0.193249 -0.139284             0
4  1.580589  1.730152 -0.860949  ...  0.071648  0.340601             0

[5 rows x 13 columns]

In [0]:
predictors = oasis.drop(columns = "GOLD_Lesions")
goldLesions = oasis["GOLD_Lesions"]

nfeatures = len(predictors.columns)
nfeatures

12

In [0]:
def create_network():
  net = models.Sequential()
  net.add(layers.Dense(units = 128, activation = 'relu', input_shape = (nfeatures, )))
  net.add(layers.Dense(units = 64, activation = 'relu'))
  net.add(layers.Dense(units = 32, activation = 'relu'))
  net.add(layers.Dense(units = 1, activation = 'sigmoid'))

  net.compile(optimizer = 'rmsprop', loss = 'binary_crossentropy', metrics = ['accuracy'])

  return net

In [0]:
neuralNetwork = KerasClassifier(build_fn = create_network, 
                                epochs = 10,
                                batch_size = 100,
                                verbose = 0)

In [0]:
accuracy = cross_val_score(neuralNetwork, predictors, goldLesions, cv = 10)
accuracy.mean()

0.8199999928474426

The cross-validated accuracy using a neural network is 0.8199999928474426

# Logistic Regression Cross Validation

In [0]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import sklearn.linear_model as lm
import statsmodels.api as sm
import sklearn.model_selection

In [0]:
oasis.head()

In [0]:
predictors = oasis.drop(columns = "GOLD_Lesions")
goldLesions = oasis["GOLD_Lesions"]

In [0]:
from sklearn.linear_model import LogisticRegression
logReg = LogisticRegression()

In [0]:
accuracyLogReg = sklearn.model_selection.cross_val_predict(logReg, predictors, goldLesions, cv = 10)
accuracyLogReg.mean()

0.51

The cross validated accuracy using a logistic regression model is 0.51. 